### PR TITLE
[Detection/Instance-Segmentation] Adaptive Training Schedule Hook

### PR DESCRIPTION
**Pull request by harimkang**
### This PR includes :
- Add Adaptive Training Hook for Updating validation interval with LR scheduler & Early-Stop scheduler
- Apply Adaptive Training Hook (validation interval only) to Instance-segmentation & Object-detection
- EMA fix (https://github.com/openvinotoolkit/model_preparation_algorithm/pull/39) merged in this PR
- Apply Adaptive EMA Momentum

### CODE REVIEW

**Comment by harimkang** [comment]
First of all, we were able to set it to the corresponding value and obtain training time in accordance with detection and instance-segmentation. However, we will experiment with additional experiments + adaptive interval in this regard.

**Comment by harimkang** [comment]
Hi @eunwoosh, This is the code for applying adaptive validation interval. This is applied in a small dataset and increases up to 5 interval. Please check if there is anything that affects HPO.

**Comment by eunwoosh** [comment]
@harimkang I checked that HPO works well with this PR.

**Review by eunwoosh — APPROVED** [review]

File: mpa/det/stage.py [hunk 1]
```diff
@@ -286,7 +286,23 @@ def configure_task_cls_incr(self, cfg, task_adapt_type, org_model_classes, model
...
+            adaptive_ema = cfg.get('adaptive_ema', False)
+            if adaptive_ema:
+                update_or_add_custom_hook(
+                    cfg,
+                    ConfigDict(
+                        type='CustomModelEMAHook',
+                        priority='ABOVE_NORMAL',
+                        epoch_momentum=0.4
+                    )
+                )
```
Comment (goodsong81) [inline mpa/det/stage.py]: How about get epoch_momentum from config? like adaptive_ema = dict(epoch_momentum=0.4)
Also we can set default value in CustomModelEMAHook constructor.

File: mpa/det/stage.py [hunk 2]
```diff
@@ -286,7 +286,23 @@ def configure_task_cls_incr(self, cfg, task_adapt_type, org_model_classes, model
...
+            else:
+                update_or_add_custom_hook(cfg, ConfigDict(type='EMAHook', priority='ABOVE_NORMAL', momentum=0.1))
+
+            adaptive_validation_interval = cfg.get('adaptive_validation_interval', False)
+            if adaptive_validation_interval:
+                update_or_add_custom_hook(cfg, ConfigDict(type='AdaptiveTrainingHook'))
```
Comment (goodsong81) [inline mpa/det/stage.py]: Similarly, we can get some optional parameters from outside, if any.
ada_cfg = cfg.get(..., {})
if ada_cfg:
   update_or_add_custom_hook(cfg, ConfigDict(type='AdaptiveTrainingHook', **_**ada_cfg_**))

File: mpa/modules/hooks/adaptive_training_hooks.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+from mmdet.core.evaluation.eval_hooks import EvalHook
+from mpa.utils.logger import get_logger
+from mpa.modules.hooks.early_stopping_hook import EarlyStoppingHook
+
+logger = get_logger()
+
+
+@HOOKS.register_module()
+class AdaptiveTrainingHook(Hook):
+    """Adaptive Training Scheduling Hook
```
Comment (goodsong81) [inline mpa/modules/hooks/adaptive_training_hooks.py]: Name seems too general. We may want to put other adaptive controls here (including adaptive EMA). Maybe adaptive train(ing?) schedule(ing) hook?

File: mpa/modules/hooks/adaptive_training_hooks.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                        math.ceil((self.base_es_patience / adaptive_interval)),
+                        self.min_es_patience,
+                    )
+                    logger.info(f"Update Early-Stop patience: {hook.patience}")
+                elif isinstance(hook, CheckpointHook):
+                    hook.interval = adaptive_interval
+            self.initialized = True
+
+    def get_adaptive_interval(self, iter_per_epoch):
+        decay = -0.025
```
Comment (goodsong81) [inline mpa/modules/hooks/adaptive_training_hooks.py]: Maybe initializer param or class constant?

File: recipes/stages/_base_/models/detectors/ssd.custom.py [hunk 5]
```diff
@@ -7,28 +7,4 @@
...
-                dict(type='Resize', keep_ratio=False),
-                dict(type='Normalize', **__img_norm_cfg),
-                dict(type='Pad', size_divisor=32),
-                dict(type='ImageToTensor', keys=['img']),
-                dict(type='Collect', keys=['img'])
-            ],
-        ),
-    ),
-)
-
```
Comment (goodsong81) [inline recipes/stages/_base_/models/detectors/ssd.custom.py:33]: Why did you remove this part?

**Review by goodsong81 — APPROVED** [review]
Looks OK but I left some comment.
Please make some change if you think needed.

File: recipes/stages/_base_/models/detectors/ssd.custom.py (hunk 5 above)
Comment (harimkang) [reply recipes/stages/_base_/models/detectors/ssd.custom.py:33]: @jaegukhyun Could you leave an answer to this?

File: mpa/det/stage.py (hunk 1 above)
Comment (harimkang) [reply mpa/det/stage.py]: Thank you for the good review. I revised it as you suggested in the comment.

File: mpa/det/stage.py (hunk 2 above)
Comment (harimkang) [reply mpa/det/stage.py]: I revised it the same as the above correction.

File: mpa/modules/hooks/adaptive_training_hooks.py (hunk 3 above)
Comment (harimkang) [reply mpa/modules/hooks/adaptive_training_hooks.py]: Modified to 'AdaptiveTrainSchedulingHook'

File: mpa/modules/hooks/adaptive_training_hooks.py (hunk 4 above)
Comment (harimkang) [reply mpa/modules/hooks/adaptive_training_hooks.py]: The variable was moved to the initializer parameter.

File: recipes/stages/_base_/models/detectors/ssd.custom.py (hunk 5 above)
Comment (goodsong81) [reply recipes/stages/_base_/models/detectors/ssd.custom.py:33]: Maybe, it's due to separate data pipeline file in model template directories.

**Comment by goodsong81** [comment]
LGTM.
